### PR TITLE
Upgrade to OpenZeppelin V5

### PR DESCRIPTION
**Pull request by andreivladbrg**
Closes https://github.com/sablier-labs/v2-core/issues/702.

Notes about PR:
- [OpenZeppelin V5](https://github.com/OpenZeppelin/openzeppelin-contracts/releases/tag/v5.0.0) removes `_after` and `_before` hooks, and implements a `_update` one
- some of the functions and parameters have been renamed
- since `ERC20` and `ERC721` are now abtract contracts, mocks were required. i know that `forge-std` offers an implementation, but they don't implemented the OpenZeppelin version (which we need in our code for the hooks called) and it would not be consistent. also, we would have needed to upgrade to the a newer version, which would trigger other incompatibilities. thus, it is better to keep the mocks in our code base.
- OpenZeppelin uses custom errors now

<details>
    <summary> Snip from release </summary>

image

</details>

### CODE REVIEW

**Review by smol-ninja — CHANGES_REQUESTED** [review]

File: src/abstracts/SablierV2Lockup.sol [hunk 1]
```diff
@@ -74,7 +74,7 @@ abstract contract SablierV2Lockup is
     /// @inheritdoc ISablierV2Lockup
     function getRecipient(uint256 streamId) external view override returns (address recipient) {
         // Checks: the stream NFT exists.
-        _requireMinted({ tokenId: streamId });
+        _requireOwned({ tokenId: streamId });
```
[the remark this answers is not in the document]
Comment (andreivladbrg) [reply src/abstracts/SablierV2Lockup.sol:77]: correct, missed this

File: src/abstracts/SablierV2Lockup.sol [hunk 2]
```diff
@@ -403,6 +370,35 @@ abstract contract SablierV2Lockup is
...
 
+    /// @notice Overrides the internal ERC-721 `_update` function to check that the stream is transferable and emit
+    /// an ERC-4906 event.
+    /// @dev There are two cases when the transferable flag is ignored:
+    /// - If `from` is 0, then the update is a mint and is allowed.
+    /// - If `to` is 0, then the update is a burn and is also allowed.
+    /// @param to The address of the new recipient of the stream.
+    /// @param streamId Id of the stream to update.
+    /// @param auth Optional parameter. If the value is non 0, the upstream implementation of this function will check
+    /// that `auth` is either the recipient of the stream, or approved to operate on the stream (by the recipient).
```
[the remark this answers is not in the document]
Comment (andreivladbrg) [reply src/abstracts/SablierV2Lockup.sol:381]: fair

File: src/abstracts/SablierV2Lockup.sol [hunk 3]
```diff
@@ -403,6 +370,35 @@ abstract contract SablierV2Lockup is
     /// @dev Retrieves the stream's status without performing a null check.
     function _statusOf(uint256 streamId) internal view virtual returns (Lockup.Status);
 
+    /// @notice Overrides the internal ERC-721 `_update` function to check that the stream is transferable and emit
+    /// an ERC-4906 event.
```
[the remark this answers is not in the document]
Comment (andreivladbrg) [reply src/abstracts/SablierV2Lockup.sol:374]: i highly suggest leaving it in the `notice`, as it is an essential aspect of what the function does and not just a detail

[the remark this answers is not in the document]
Comment (smol-ninja) [reply src/abstracts/SablierV2Lockup.sol:371]: OK. Fair enough.

**Review by smol-ninja — APPROVED** [review]

**Comment by PaulRBerg** [comment]
I will review this once I get to review all of your recent work (in a batched review session)

**Comment by andreivladbrg** [comment]
> I will review this once I get to review all of your recent work

fyi: you've reviewed this one time before: https://github.com/sablier-labs/v2-core/pull/703

i just rebased from `staging`

**Comment by PaulRBerg** [comment]
IK, wanted to review again

File: script/Init.s.sol [hunk 4]
```diff
@@ -14,7 +14,7 @@ import { Broker, LockupDynamic, LockupLinear } from "../src/types/DataTypes.sol"
 import { BaseScript } from "./Base.s.sol";
 
 interface IERC20Mint {
-    function mint(address beneficiary, uint256 amount) external;
+    function mint(address beneficiary, uint256 value) external;
```
Comment (PaulRBerg) [inline script/Init.s.sol:17]: Why did you change the `amount` to `value` here?

**Comment by PaulRBerg** [comment]
@andreivladbrg can we please get an answer from you here

**Comment by andreivladbrg** [comment]
Sorry for not responding to this earlier. I did not receive an email notification for the first message.

The mock contracts in `forge-std` are replicas of solmate's `ERC20` & `ERC721`, the reason being that they are very minimalist.

However, since we implement the `ERC721._update` [hook](https://github.com/sablier-labs/v2-core/blob/76e29fee33befaeb54a82473c28cf182e6f8855d/src/abstracts/SablierV2Lockup.sol#L522), which is expected to be called in multiple functions (transfer, mint, burn, etc.) in OpenZeppelin version, solmate does not have this implementation. Therefore, a lot of the tests would fail because of this.

For the `ERC20`, I guess we could switch to the `forge-std` version, but we must test it to see if it works.

Does this answer your question?

**Comment by andreivladbrg** [comment]
Regarding the `ERC20`, the solmate version does not inherit an interface `IERC20`, and we would need to refactor our tests to take as parameter an `address` instead of `IERC20`. Also, the functions have different named params, so it would be different from `src` to `test`.

Do you consider that it would be that needed to remove the mocks?

I personally don't think we should make this change atm.

File: script/Init.s.sol (hunk 4 above)
Comment (andreivladbrg) [reply script/Init.s.sol:17]: because OpenZeppelin does not say `amount` anymore but `value` for all function parameter names

so, i wanted to follow their approach

**Comment by PaulRBerg** [comment]
> Does this answer your question?

Yes, thank you for answering. Let's proceed with using their ERC-721 mock.

> Do you consider that it would be needed to remove the ERC20 mock? @PaulRBerg

No. Let's keep using our mock. It would be nice if OpenZeppelin offered a mock for this, though ..
